langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java

```diff
@@ -116,18 +122,32 @@ private JsonNode executeAndWait(Request request, Long id) {
 
     private EventSource startSseChannel(boolean logResponses) {
         Request request = new Request.Builder().url(sseUrl).build();
-        CompletableFuture<Void> initializationFinished = new CompletableFuture<>();
+        CompletableFuture<String> initializationFinished = new CompletableFuture<>();
         SseEventListener listener = new SseEventListener(pendingOperations, logResponses, initializationFinished);
         EventSource eventSource = EventSources.createFactory(client).newEventSource(request, listener);
-        // wait for the SSE channel to be created, throw an exception if that failed
+        // wait for the SSE channel to be created, receive the POST url from the server, throw an exception if that
+        // failed
         try {
-            initializationFinished.get();
+            int timeout = client.callTimeoutMillis() > 0 ? client.callTimeoutMillis() : Integer.MAX_VALUE;
+            String relativePostUrl = initializationFinished.get(timeout, TimeUnit.MILLISECONDS);
+            postUrl = buildAbsolutePostUrl(relativePostUrl);
+            log.debug("Received the server's POST URL: " + postUrl);
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
         return eventSource;
     }
 
+    private String buildAbsolutePostUrl(String relativePostUrl) {
+        try {
+            URL sseUrl = new URL(this.sseUrl);
```

Comment:
Perharps we should use `URI` here instead?

Reply:
Like this?
```
URI sseUrl = new URI(this.sseUrl);
URL postUrl = new URL(sseUrl.toURL(), relativePostUrl);
return postUrl.toString();
```

or what else do you have in mind?

Reply:
I have heard bad things about `URL` (how it initiates a connection in `equals()`, etc.).
IIUC, the general advice is to use `URI` instead.
So if this is true, I assume we should use URI everywhere?

Reply:
Ok, I've replaced it with `URI.create(this.sseUrl).resolve(relativePostUrl).toString()`

Reply:
Thanks!